libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/PlannedRouteRefreshController.kt

```diff
@@ -0,0 +1,148 @@
+    private val routeRefresherExecutor: RouteRefresherExecutor,
+    private val routeRefreshOptions: RouteRefreshOptions,
+    private val stateHolder: RouteRefreshStateHolder,
+    private val listener: RouteRefresherListener,
+    private val cancellableHandler: CancellableHandler,
+    private val retryStrategy: RetryRouteRefreshStrategy,
+) {
+
+    constructor(
+        routeRefresherExecutor: RouteRefresherExecutor,
+        routeRefreshOptions: RouteRefreshOptions,
+        stateHolder: RouteRefreshStateHolder,
+        scope: CoroutineScope,
+        listener: RouteRefresherListener,
+    ) : this(
+        routeRefresherExecutor,
+        routeRefreshOptions,
+        stateHolder,
+        listener,
+        CancellableHandler(scope),
+        RetryRouteRefreshStrategy(maxRetryCount = MAX_RETRY_COUNT)
+    )
+
+    private var paused = false
+    var routesToRefresh: List<NavigationRoute>? = null
+        private set
+
+    fun startRoutesRefreshing(routes: List<NavigationRoute>) {
+        cancellableHandler.cancelAll()
+        routesToRefresh = null
+        if (routes.isEmpty()) {
+            logI("Routes are empty", RouteRefreshLog.LOG_CATEGORY)
+            stateHolder.reset()
+            return
+        }
+        val routesValidationResults = routes.map { RouteRefreshValidator.validateRoute(it) }
+        if (
+            routesValidationResults.any { it is RouteRefreshValidator.RouteValidationResult.Valid }
+        ) {
+            routesToRefresh = routes
+            scheduleNewUpdate(routes)
+        } else {
+            val message =
+                RouteRefreshValidator.joinValidationErrorMessages(
+                    routesValidationResults.mapIndexed { index, routeValidationResult ->
+                        routeValidationResult to routes[index]
+                    }
+                )
+            val logMessage = "No routes which could be refreshed. $message"
+            logI(logMessage, RouteRefreshLog.LOG_CATEGORY)
+            stateHolder.onFailure(logMessage)
+            stateHolder.reset()
+        }
+    }
+
+    fun pause() {
+        if (!paused) {
+            paused = true
+            cancellableHandler.cancelAll()
+        }
+    }
+
+    fun resume() {
+        if (paused) {
+            paused = false
+            routesToRefresh?.let {
+                if (retryStrategy.shouldRetry()) {
+                    scheduleUpdateRetry(it, shouldNotifyOnStart = true)
+                }
+            }
+        }
+    }
+
+    private fun scheduleNewUpdate(routes: List<NavigationRoute>) {
+        retryStrategy.reset()
+        postAttempt {
+            executePlannedRefresh(routes, shouldNotifyOnStart = true)
+        }
+    }
+
+    private fun scheduleUpdateRetry(routes: List<NavigationRoute>, shouldNotifyOnStart: Boolean) {
+        postAttempt { executePlannedRefresh(routes, shouldNotifyOnStart = shouldNotifyOnStart) }
+    }
```

Comment:
```suggestion
    private fun scheduleUpdateRetry(routes: List<NavigationRoute>, shouldNotifyOnStart: Boolean) {                
        retryStrategy.onNextAttempt()
        postAttempt { executePlannedRefresh(routes, shouldNotifyOnStart = shouldNotifyOnStart) }
    }
```
this looks clearer to me